File: mortie/greedy_polygon.py
"""
Greedy polygon subdivision for morton indices.

This module provides functions to generate compact morton index coverage
for arbitrary geometries using greedy subdivision with configurable constraints.
"""

import numpy as np


def greedy_morton_polygon(lat, lon, order=18, max_boxes=16, ordermax=None, verbose=False):
    """
    Calculate compact morton index coverage using greedy subdivision.

    This function finds the highest-order morton indices that fully contain
    all points in the input geometry. It greedily subdivides until reaching
    the max_boxes ceiling or ordermax limit.

    Parameters
    ----------
    lat : array-like
        Latitude values in degrees
    lon : array-like
        Longitude values in degrees
    order : int, default=18
        Morton cell order to start with (higher = finer resolution)
    max_boxes : int, default=16
        Maximum number of morton boxes allowed
    ordermax : int, optional
        Maximum order (resolution) for subdivision. Branches that reach
        this order will not be subdivided further. If None, no order limit.
    verbose : bool, default=False
        Print subdivision progress

    Returns
    -------
    mbbox : ndarray
        Array of morton indices that form the minimum bounding box
    mbbox_orders : ndarray
        Array of orders for each morton index

    Examples
    --------
    >>> import numpy as np
    >>> lat = np.array([-75, -75, -70, -70])
    >>> lon = np.array([-80, -70, -70, -80])
    >>> mbbox, orders = greedy_morton_polygon(lat, lon, max_boxes=16, ordermax=6)

    Notes
    -----
    The algorithm uses a greedy strategy that prioritizes splitting coarser
    (lower-order) boxes first to maintain balanced spatial coverage across
    the geometry. This prevents over-subdivision of any single region.
    """
    from . import geo2mort

    # Convert to numpy arrays
    lat = np.asarray(lat)
    lon = np.asarray(lon)

    # Remove NaN values
    valid = ~(np.isnan(lat) | np.isnan(lon))
    lat = lat[valid]
    lon = lon[valid]

    if len(lat) == 0:
        raise ValueError("No valid points provided")

    # Calculate morton indices at specified order
    midx = geo2mort(lat, lon, order=order)

    # Convert to string array for character-by-character analysis
    midx_str = np.array([str(m) for m in midx])

    if verbose:
        print(f"Starting greedy subdivision with max_boxes={max_boxes}, ordermax={ordermax}")
        print(f"Total morton indices at order {order}: {len(midx_str):,}\n")

    # Greedy subdivision
    morton_boxes = _greedy_subdivide(midx_str, max_boxes, ordermax=ordermax, verbose=verbose)

    if verbose:
        print(f"\n{'='*70}")
        print(f"Final result: {len(morton_boxes)} morton boxes")
        print(f"{'='*70}\n")

    # Convert strings back to integers and calculate orders
    morton_indices = np.array([int(box) for box in morton_boxes])
    morton_orders = np.array([_estimate_order_from_prefix(box) for box in morton_boxes])

    return morton_indices, morton_orders


def _greedy_subdivide(morton_strings, max_boxes, ordermax=None, verbose=False):
    """
    Greedily subdivide morton indices until hitting limits.

    Parameters
    ----------
    morton_strings : array-like of str
        Morton indices as strings
    max_boxes : int
        Maximum number of boxes allowed
    ordermax : int, optional
        Maximum order for subdivision
    verbose : bool
        Print progress

    Returns
    -------
    boxes : list of str
        List of morton index prefixes
    """
    # Start with one box containing all indices
    boxes = [{'prefix': _find_common_prefix(morton_strings), 'indices': morton_strings}]

    if verbose:
        print(f"Initial: 1 box covering {len(morton_strings):,} indices")
        print(f"  Box 1: prefix='{boxes[0]['prefix']}'\n")

    iteration = 0
    while len(boxes) < max_boxes:
        iteration += 1
        if verbose:
            print(f"Iteration {iteration}: Current boxes: {len(boxes)}/{max_boxes}")

        # Find which box can be split
        best_box_idx = None
        best_split = None
        max_split_count = 0
        best_order = float('inf')

        for i, box in enumerate(boxes):
            # Check if this box has reached ordermax
            if ordermax is not None:
                box_order = _estimate_order_from_prefix(box['prefix'])
                if box_order >= ordermax:
                    if verbose and iteration == 1:
                        print(f"  Box {i+1} ('{box['prefix']}') at order {box_order} >= ordermax {ordermax}, skipping")
                    continue

            # Try to split this box
            split_result = _try_split(box['prefix'], box['indices'])

            if split_result is not None:
                n_splits = len(split_result)
                new_total = len(boxes) - 1 + n_splits

                if new_total <= max_boxes and n_splits > 1:
                    # Priority: lower order first, then more splits
                    box_order = _estimate_order_from_prefix(box['prefix'])

                    if best_box_idx is None:
                        best_box_idx = i
                        best_split = split_result
                        max_split_count = n_splits
                        best_order = box_order
                    else:
                        if (box_order < best_order or
                            (box_order == best_order and n_splits > max_split_count)):
                            best_box_idx = i
                            best_split = split_result
                            max_split_count = n_splits
                            best_order = box_order

        if best_box_idx is None:
            if verbose:
                print(f"  → No more splits possible within budget\n")
            break

        # Apply the best split
        old_box = boxes[best_box_idx]
        if verbose:
            print(f"  → Splitting box {best_box_idx+1} ('{old_box['prefix']}', order {best_order}) into {len(best_split)} sub-boxes")

        boxes.pop(best_box_idx)

        # Clip children to ordermax if necessary
        if ordermax is not None:
            clipped_split = []
            for child in best_split:
                child_order = _estimate_order_from_prefix(child['prefix'])
                if child_order > ordermax:
                    # Clip prefix to ordermax
                    clipped_prefix = child['prefix'][:ordermax + 2]  # +2 for minus sign and parent digit
                    if verbose:
                        print(f"    Clipping child '{child['prefix']}' (order {child_order}) → '{clipped_prefix}' (order {ordermax})")
                    child['prefix'] = clipped_prefix
                clipped_split.append(child)
            best_split = clipped_split

        boxes.extend(best_split)

        if verbose:
            print(f"  → New total: {len(boxes)} boxes\n")

    # Extract prefixes
    result = [box['prefix'] for box in boxes]

    if verbose:
        print(f"Final subdivision complete:")
        for i, box in enumerate(boxes, 1):
            print(f"  Box {i:2d}: '{box['prefix']}' ({len(box['indices']):,} indices)")

    return result


def _try_split(prefix, indices):
    """Try to split a box into sub-boxes."""
    if len(np.unique(indices)) == 1:
        return None

    divergence_pos = len(prefix)

    if divergence_pos >= min(len(s) for s in indices):
        return None

    next_chars = np.array([s[divergence_pos] if divergence_pos < len(s) else ''
                           for s in indices])
    unique_next_chars = np.unique(next_chars[next_chars != ''])

    if len(unique_next_chars) <= 1:
        return None

    # Create sub-boxes
    sub_boxes = []
    for char in unique_next_chars:
        mask = np.array([s[divergence_pos] == char if divergence_pos < len(s) else False
                        for s in indices])
        group_indices = indices[mask]

        if len(group_indices) > 0:
            group_prefix = _find_common_prefix(group_indices)
            sub_boxes.append({'prefix': group_prefix, 'indices': group_indices})

    return sub_boxes if len(sub_boxes) > 1 else None


def _estimate_order_from_prefix(prefix):
    """
    Estimate morton order from prefix string.

    Parameters
    ----------
    prefix : str
        Morton index prefix

    Returns
    -------
    order : int
        Estimated order (number of digits - 1)
    """
    digits = prefix.lstrip('-')
    return len(digits) - 1


def _find_common_prefix(strings):
    """
    Find longest common prefix among strings.

    Parameters
    ----------
    strings : array-like of str
        Array of strings

    Returns
    -------
    prefix : str
        Longest common prefix
    """
    if len(strings) == 0:
        return ""

    if len(strings) == 1:
        return strings[0]

    prefix = strings[0]

    for i in range(len(prefix)):
        char = prefix[i]
        for s in strings[1:]:
            if i >= len(s) or s[i] != char:
                return prefix[:i]

    return prefix

File: src_rust/src/greedy_polygon.rs
//! Greedy polygon subdivision for morton indices
//!
//! This module provides a greedy algorithm to generate compact morton index
//! coverage for arbitrary geometries with configurable constraints.

use std::collections::HashSet;

/// A box in the subdivision tree
#[derive(Clone)]
struct MortonBox {
    prefix: String,
    indices: Vec<String>,
}

/// Estimate order from a morton prefix string
fn estimate_order_from_prefix(prefix: &str) -> usize {
    // Remove minus sign and count digits
    let digits = prefix.trim_start_matches('-');
    if digits.is_empty() {
        return 0;
    }
    digits.len() - 1
}

/// Find longest common prefix among strings
fn find_common_prefix(strings: &[String]) -> String {
    if strings.is_empty() {
        return String::new();
    }

    if strings.len() == 1 {
        return strings[0].clone();
    }

    let first = &strings[0];
    let mut prefix_len = 0;

    for (i, ch) in first.chars().enumerate() {
        // Check if all strings have same character at position i
        let all_match = strings[1..].iter().all(|s| {
            s.chars().nth(i).map_or(false, |c| c == ch)
        });

        if all_match {
            prefix_len = i + 1;
        } else {
            break;
        }
    }

    first.chars().take(prefix_len).collect()
}

/// Try to split a box into sub-boxes
fn try_split(prefix: &str, indices: &[String]) -> Option<Vec<MortonBox>> {
    // Check if all indices are identical
    let unique: HashSet<&String> = indices.iter().collect();
    if unique.len() == 1 {
        return None;
    }

    let divergence_pos = prefix.len();

    // Check if we've exhausted string length
    let min_len = indices.iter().map(|s| s.len()).min().unwrap_or(0);
    if divergence_pos >= min_len {
        return None;
    }

    // Get unique characters at divergence position
    let mut unique_chars: Vec<char> = indices
        .iter()
        .filter_map(|s| s.chars().nth(divergence_pos))
        .collect::<HashSet<_>>()
        .into_iter()
        .collect();

    if unique_chars.len() <= 1 {
        return None;
    }

    unique_chars.sort();

    // Create sub-boxes
    let mut sub_boxes = Vec::new();

    for ch in unique_chars {
        let group_indices: Vec<String> = indices
            .iter()
            .filter(|s| s.chars().nth(divergence_pos) == Some(ch))
            .cloned()
            .collect();

        if !group_indices.is_empty() {
            let group_prefix = find_common_prefix(&group_indices);
            sub_boxes.push(MortonBox {
                prefix: group_prefix,
                indices: group_indices,
            });
        }
    }

    if sub_boxes.len() > 1 {
        Some(sub_boxes)
    } else {
        None
    }
}

/// Greedy subdivision of morton indices
///
/// # Arguments
/// * `morton_strings` - Morton indices as strings
/// * `max_boxes` - Maximum number of boxes allowed
/// * `ordermax` - Optional maximum order for subdivision
///
/// # Returns
/// Vector of morton index prefix strings
pub fn greedy_subdivide(
    morton_strings: Vec<String>,
    max_boxes: usize,
    ordermax: Option<usize>,
) -> Vec<String> {
    // Start with one box containing all indices
    let mut boxes = vec![MortonBox {
        prefix: find_common_prefix(&morton_strings),
        indices: morton_strings,
    }];

    while boxes.len() < max_boxes {

        // Find which box can be split
        let mut best_box_idx: Option<usize> = None;
        let mut best_split: Option<Vec<MortonBox>> = None;
        let mut max_split_count = 0;
        let mut best_order = usize::MAX;

        for (i, box_item) in boxes.iter().enumerate() {
            // Check if this box has reached ordermax
            if let Some(om) = ordermax {
                let box_order = estimate_order_from_prefix(&box_item.prefix);
                if box_order >= om {
                    continue;
                }
            }

            // Try to split this box
            if let Some(split_result) = try_split(&box_item.prefix, &box_item.indices) {
                let n_splits = split_result.len();
                let new_total = boxes.len() - 1 + n_splits;

                if new_total <= max_boxes && n_splits > 1 {
                    // Priority: lower order first, then more splits
                    let box_order = estimate_order_from_prefix(&box_item.prefix);

                    if best_box_idx.is_none()
                        || box_order < best_order
                        || (box_order == best_order && n_splits > max_split_count)
                    {
                        best_box_idx = Some(i);
                        best_split = Some(split_result);
                        max_split_count = n_splits;
                        best_order = box_order;
                    }
                }
            }
        }

        // If no splits possible, break
        let Some(idx) = best_box_idx else {
            break;
        };

        // Apply the best split
        let mut new_split = best_split.unwrap();

        // Clip children to ordermax if necessary
        if let Some(om) = ordermax {
            for child in &mut new_split {
                let child_order = estimate_order_from_prefix(&child.prefix);
                if child_order > om {
                    // Clip prefix to ordermax
                    // ordermax + 2 for minus sign and parent digit
                    let target_len = om + 2;
                    if child.prefix.len() > target_len {
                        child.prefix.truncate(target_len);
                    }
                }
            }
        }

        // Remove old box and add new sub-boxes
        boxes.remove(idx);
        boxes.extend(new_split);
    }

    // Extract prefixes
    boxes.iter().map(|b| b.prefix.clone()).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_find_common_prefix_empty() {
        let strings = vec![];
        assert_eq!(find_common_prefix(&strings), "");
    }

    #[test]
    fn test_find_common_prefix_single() {
        let strings = vec!["-511".to_string()];
        assert_eq!(find_common_prefix(&strings), "-511");
    }

    #[test]
    fn test_find_common_prefix_multiple() {
        let strings = vec!["-511111".to_string(), "-511222".to_string(), "-511333".to_string()];
        assert_eq!(find_common_prefix(&strings), "-511");
    }

    #[test]
    fn test_estimate_order() {
        assert_eq!(estimate_order_from_prefix("-6"), 0);
        assert_eq!(estimate_order_from_prefix("-61"), 1);
        assert_eq!(estimate_order_from_prefix("-611"), 2);
        assert_eq!(estimate_order_from_prefix("-6111"), 3);
    }

    #[test]
    fn test_greedy_subdivide_basic() {
        let morton_strings = vec![
            "-311111".to_string(),
            "-311222".to_string(),
            "-411111".to_string(),
            "-411222".to_string(),
        ];

        let result = greedy_subdivide(morton_strings, 10, None);
        assert!(result.len() >= 2);
        assert!(result.len() <= 10);
    }

    #[test]
    fn test_greedy_subdivide_respects_max_boxes() {
        let morton_strings = vec![
            "-311111".to_string(),
            "-322222".to_string(),
            "-411111".to_string(),
            "-422222".to_string(),
        ];

        let result = greedy_subdivide(morton_strings, 3, None);
        assert!(result.len() <= 3);
    }

    #[test]
    fn test_greedy_subdivide_ordermax() {
        let morton_strings = vec![
            "-3111111111".to_string(),
            "-3222222222".to_string(),
        ];

        let result = greedy_subdivide(morton_strings, 10, Some(2));

        // All results should be order <= 2 (3 digits or less)
        for prefix in &result {
            let order = estimate_order_from_prefix(prefix);
            assert!(order <= 2, "Order {} exceeds ordermax 2 for prefix {}", order, prefix);
        }
    }
}
